File: Problems/Algorithms/735. Asteroid Collision/AsteroidCollision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        
        for (int i = 0; i < asteroids.length; i++) {
            int curr = asteroids[i];
            
            boolean flag = false;
            while (!stack.empty() && curr < 0 && stack.peek() > 0) {
                int prev = stack.pop();
                if (Math.abs(curr) < Math.abs(prev)) {
                    curr = prev;
                } else if (curr + prev == 0) {
                    flag = true;
                    break;
                }
            }
            
            if (!flag) stack.push(curr);
        }
        
        int size = stack.size();
        int[] ans = new int[size];
        for (int i = size-1; i >= 0; i--) {
            ans[i] = stack.pop();
        }
        return ans;
    }
}

File: Problems/Algorithms/735. Asteroid Collision/asteroid-collision.cpp
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> stack;

        for (int asteroid : asteroids) {
            int flag = 1;
            while (!stack.empty() && (stack.top() > 0 && asteroid < 0)) {
                if (abs(stack.top()) < abs(asteroid)) {
                    stack.pop();
                    continue;
                } else if (abs(stack.top()) == abs(asteroid)) {
                    stack.pop();
                }

                flag = 0;
                break;
            }

            if (flag) {
                stack.push(asteroid);
            }
        }

        vector<int> results(stack.size());
        for (int i = results.size() - 1; i >= 0; i--) {
            results[i] = stack.top();
            stack.pop();
        }

        return results;
    }
};
